File: src/challenges/python/problems/simplify_path_test.py
# https://leetcode.com/problems/simplify-path

from core.problem_base import *

class SimplifyPath(ProblemBase):
    def Solution(self, path: str) -> str:
        stack = []
        segment = ""

        for i in range(len(path)):
            if path[i] != '/':
                segment += path[i]

            if path[i] == '/' or i == len(path) - 1:
                dir = segment

                if dir == ".." and len(stack) > 0:
                    stack.pop()
                elif len(dir) > 0 and dir != "." and dir != "..":
                    stack.append(dir)

                segment = ""

        builder = ""
        if len(stack) <= 0:
            builder = "/"

        while len(stack) > 0:
            sub = stack.pop()
            builder = f"/{sub}" + builder

        return builder

if __name__ == '__main__':
    TestGen(SimplifyPath) \
        .Add(lambda tc: tc.Param("/home/").Result("/home")) \
        .Add(lambda tc: tc.Param("/../").Result("/")) \
        .Add(lambda tc: tc.Param("/home//foo/").Result("/home/foo")) \
        .Add(lambda tc: tc.Param("/home/../foo/").Result("/foo")) \
        .Add(lambda tc: tc.Param("/..hidden").Result("/..hidden")) \
        .Add(lambda tc: tc.Param("/a/./b/../../c/").Result("/a/./b/../../c/")) \
        .Run()

File: src/challenges/go/problems/simplify_path_test.go
// https://leetcode.com/problems/simplify-path

package problems

import (
	"strings"
	"testing"

	"github.com/dobermanch/leetcode/core"
)

type SimplifyPath struct{}

func TestSimplifyPath(t *testing.T) {
	gen := core.TestSuite[SimplifyPath]{}
	gen.Add(func(tc *core.TestCase) {
		tc.Param("/home//foo/").Result("/home/foo")
	}).Add(func(tc *core.TestCase) {
		tc.Param("/home/").Result("/home")
	}).Add(func(tc *core.TestCase) {
		tc.Param("/../").Result("/")
	}).Run(t)
}

func (SimplifyPath) Solution(path string) string {
	stack := []string{}
	segment := new(strings.Builder)

	for i := 0; i < len(path); i++ {
		if path[i] != '/' {
			segment.WriteByte(path[i])
		}

		if path[i] == '/' || i == len(path)-1 {
			dir := segment.String()
			if dir == ".." && len(stack) > 0 {
				stack = stack[:len(stack)-1]
			} else if len(dir) > 0 && dir != "." && dir != ".." {
				stack = append(stack, dir)
			}

			segment.Reset()
		}
	}

	builder := new(strings.Builder)
	if len(stack) <= 0 {
		builder.WriteByte('/')
	}

	for i := 0; i < len(stack); i++ {
		builder.WriteString("/")
		builder.WriteString(stack[i])
	}

	return builder.String()
}
